### PR TITLE
 Update the rules for parsing subjects

### LINKED ISSUE #111: Parsing Epub subjects
[Epub Parsing Guide](https://github.com/readium/architecture/blob/master/streamer/parser/metadata.md#subjects) is currently too simplistic about Epub subject parsing.

> 1. if there is a one single `<dc:subject>` element, make sure keywords are not separated using commas or semicolons;
    1. if it doesn’t, the string is the value;
    2. if it does, split the string to build an array;
> 2. if there are more than one `<dc:subject>` elements, build an array using their values.

Both Readium manifest and Epub supports sortAs and localized strings in subjects, so the algorithm should be modified. I suggest the following:

1. if there is a one single `<dc:subject>` element without any sorting information or translation (possibly not compliant case), make sure keywords are not separated using commas or semicolons;
    1. if it doesn’t, the string is the value;
    2. if it does, split the string to build an array;
2. else, assume conformance to Epub spec and build an array using their values.

### PR DESCRIPTION
**Pull request by qnga**
- Add rules for `sortAs`, `code` and `scheme` properties
- Don't try to split subjects if they are refined by some `meta` element
- Make the parsing of subjects compliant with the rule 
> when parsing a publication in the streamer we always use the most complex form for each metadata to harmonize our output.

Fix #111.

### CODE REVIEW

**Comment by HadrienGardeur** [comment]
This should be reviewed by @JayPanoz since he wrote the initial document.

File: streamer/parser/metadata.md [hunk 1]
```diff
@@ -219,16 +211,24 @@ The string is the value of the `meta` element whose `property` attribute has the
 
 ## Subjects
 
-The `subject` of a publication is a key whose value is string or an array.
+The `subject` of a publication is a key whose value is, in the most complex form, an array of `subject` objects.
+
+Although each subject should have its own `dc:subject` element, this is not necessarily the case in practice, authors and authoring tools often separating multiple subjects using commas or semicolons in the same element.
```
Comment (JayPanoz) [inline streamer/parser/metadata.md]:
```suggestion
Although each subject should have its own `<dc:subject>` element, this is not necessarily the case in practice, authors and authoring tools often separating multiple subjects using commas or semicolons in the same element.
```

File: streamer/parser/metadata.md [hunk 2]
```diff
@@ -219,16 +211,24 @@ The string is the value of the `meta` element whose `property` attribute has the
...
+So, if there is a single `dc:subject` that is not refined by any property, split its content at every comma and semicolon and consider you have several `dc:subject` with shared attributes.
+
+Parse each `dc:subject` element as a [localized string](#localized-strings) to compute a language map for the subject's `name`.
```
Comment (JayPanoz) [inline streamer/parser/metadata.md]:
```suggestion
Parse each `<dc:subject>` element as a [localized string](#localized-strings) to compute a language map for the subject’s `name`.
```

File: streamer/parser/metadata.md [hunk 3]
```diff
@@ -111,52 +105,56 @@ The contributor object may also contain a `sortAs` string, used to sort the cont
...
+| dc:contributor | ill                             | illustrator |
+| dc:contributor | art                             | artist      |
+| dc:contributor | clr                             | colorist    |
+| dc:contributor | nrt                             | narrator    |
+| dc:contributor | \<empty\> or \<unknown\>        | contributor |
 
 Where `opf:role` is the value of the attribute of the `<dc:element>`.
 
-The `name` of the contributor is the value of the element.
+In case of conflict, `opf:role` overrides the XML element used. So, for example, map a `<dc:publisher>` or `<dc:contributor>` element with `opf:role` aut to an author.
```
[the remark this answers is not in the document]
Comment (JayPanoz) [reply streamer/parser/metadata.md]: For clarification, is this ad hoc or is there a section of a spec that could be linked?

File: streamer/parser/metadata.md [hunk 4]
```diff
@@ -111,52 +105,56 @@ The contributor object may also contain a `sortAs` string, used to sort the cont
...
+
+Parse the carrying element as a [localized string](#localized-strings) to compute a language map for the contributor's name.
```
Comment (JayPanoz) [inline streamer/parser/metadata.md]:
```suggestion
Parse the carrying element as a [localized string](#localized-strings) to compute a language map for the contributor’s name.
```

File: streamer/parser/metadata.md [hunk 5]
```diff
@@ -111,52 +105,56 @@ The contributor object may also contain a `sortAs` string, used to sort the cont
...
+
 
 Where `role` is the value of the refine whose `scheme` is a value of `marc:relators`.
 
-To handle the `name` of the contributor:
+In case of conflict, `role` overrides the XML element used. So, for example, map a `<dc:publisher>` or `<dc:contributor>` element with `role` aut to an author.
 
-1. check if there is a refine whose propery is `alternate-script` and its corresponding `xml:lang` value;
-2. if there is none, use the value of the `<dc:element>`.
+Parse the `contributor` element as a [localized string](#localized-strings) to compute a language map for the contributor's name.
```
Comment (JayPanoz) [inline streamer/parser/metadata.md]:
```suggestion
Parse the `contributor` element as a [localized string](#localized-strings) to compute a language map for the contributor’s name.
```

File: streamer/parser/metadata.md [hunk 6]
```diff
@@ -15,51 +26,34 @@ In addition to `title`, a publication may also contain a `sortAs` string, used t
 When parsing an EPUB, we need to establish:
 
 * which title is the primary one
-* the language(s) used to express the primary title along with the associated strings
 * the string used to sort the title of the publication
-* the subtitle of the publication
-* the default language for metadata
+* a language map of the representations of the title
+* which title is the subtitle
+* a language map of the representations of the title
```
[the remark this answers is not in the document]
Comment (qnga) [reply streamer/parser/metadata.md]: That's right, thanks.

File: streamer/parser/metadata.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (qnga) [reply streamer/parser/metadata.md]: This is the spec, isn't it? If that's what you're asking, this is not part of the Epub specification.

**Comment by qnga** [comment]
Thanks for the improvements JayPanoz.

File: streamer/parser/metadata.md [hunk 7]
```diff
@@ -111,52 +105,56 @@ The contributor object may also contain a `sortAs` string, used to sort the cont
...
+| dc:contributor | ill                             | illustrator |
+| dc:contributor | art                             | artist      |
+| dc:contributor | clr                             | colorist    |
+| dc:contributor | nrt                             | narrator    |
+| dc:contributor | \<empty\> or \<unknown\>        | contributor |
 
 Where `opf:role` is the value of the attribute of the `<dc:element>`.
 
-The `name` of the contributor is the value of the element.
+In case of conflict, `opf:role` overrides the XML element used. For example, map a `<dc:publisher>` or `<dc:contributor>` element whose value for `opf:role` is `aut` to an author.
```
Comment (JayPanoz) [inline streamer/parser/metadata.md]: Alright, so I now know why that made me pause yesterday. 

I would rephrase the “case of conflict” into a case of extension/augmentation since:

> The [DC11] contributor element is used to represent the name of a person, organization, etc. that played a secondary role in the creation of the content of an EPUB Publication. ([source](https://www.w3.org/publishing/epub3/epub-packages.html#sec-opf-dccontributor))

So in theory, `<dc:contributor> + aut` should be something of a secondary author. But maybe I’m wrong about that, please feel free to chime in.

Then `role` does not apply to `<dc:publisher>`, cf. 

> Extends: contributor, creator

in [the EPUB 3.2 spec](https://www.w3.org/publishing/epub3/epub-packages.html#sec-role). Same in [the EPUB 2.0.1 spec](http://idpf.org/epub/20/spec/OPF_2.0.1_draft.htm#Section1.3.3): “the OPF role attribute added to the Dublin Core `creator` and `contributor` elements”.

File: streamer/parser/metadata.md [hunk 8]
```diff
@@ -111,52 +105,56 @@ The contributor object may also contain a `sortAs` string, used to sort the cont
...
+| dc:contributor | clr                             | colorist    |
+| dc:contributor | nrt                             | narrator    |
+| media:narrator | nrt or \<empty\> or \<unknown\> | narrator    |
+| dc:contributor | \<empty\> or \<unknown\>        | contributor |
+
 
 Where `role` is the value of the refine whose `scheme` is a value of `marc:relators`.
 
-To handle the `name` of the contributor:
+In case of conflict, `role` overrides the XML element used. So, for example, map a `<dc:publisher>` or `<dc:contributor>` element whose value for `role` is `aut` to an author.
```
Comment (JayPanoz) [inline streamer/parser/metadata.md]: Same as above.

File: streamer/parser/metadata.md (hunk 7 above)
Comment (qnga) [reply streamer/parser/metadata.md]: > So in theory, <dc:contributor> + aut should be something of a secondary author. But maybe I’m wrong about that, please feel free to chime in.

If it is so, maybe our rule is not founded and we should map <dc:contributor> + aut  to a contributor...

> Then role does not apply to <dc:publisher>, cf.

You're right, indeed.

Comment (JayPanoz) [reply streamer/parser/metadata.md]: Given some producers may put several authors in a single `<dc:creator>` so that both/all names are displayed in the UI, I’d imagine that if you can find `<dc:contributor> + aut`, there’s an underlying intent to make it secondary, and it’s a nuance to probably take into account.

But then I guess it could be useful to check if others have any thoughts? Now that we are fine-tuning that, I’m wondering whether `creator` and `contributor` have been handled well enough so far in this doc. I’m not sure about that, since these `roles` may apply to both.

Comment (JayPanoz) [reply streamer/parser/metadata.md]: Opened an issue here: https://github.com/readium/architecture/issues/126

Comment (HadrienGardeur) [reply streamer/parser/metadata.md]: I agree that `role` should not be applied to `<dc:publisher>`.

`<dc:creator>` and `<dc:contributor>` are unfortunately very vague, `creator` is not as specific as `author` for instance.
I think these two elements are vague enough to ignore the element and override them with `role`.

File: streamer/parser/metadata.md [hunk 9]
```diff
@@ -111,52 +105,56 @@ The contributor object may also contain a `sortAs` string, used to sort the cont
...
+|----------------|---------------------------------|-------------|
+| dc:creator     | aut or \<empty\> or \<unknown\> | author      |
+| dc:publisher   | pbl or \<empty\> or \<unknown\> | publisher   |
+| dc:contributor | trl                             | translator  |
+| dc:contributor | edt                             | editor      |
+| dc:contributor | ill                             | illustrator |
+| dc:contributor | art                             | artist      |
+| dc:contributor | clr                             | colorist    |
+| dc:contributor | nrt                             | narrator    |
+| dc:contributor | \<empty\> or \<unknown\>        | contributor |
```
Comment (JayPanoz) [inline streamer/parser/metadata.md]:
```suggestion
| element                      | opf:role                 | key         |
|------------------------------|--------------------------|-------------|
| dc:creator                   | \<empty\> or \<unknown\> | author      |
| dc:creator or dc:contributor | aut                      | author      |
| dc:contributor               | \<empty\> or \<unknown\> | contributor |
| dc:publisher                 | \<any\>                  | publisher   |
| dc:creator or dc:contributor | pbl                      | publisher   |
| dc:creator or dc:contributor | trl                      | translator  |
| dc:creator or dc:contributor | edt                      | editor      |
| dc:creator or dc:contributor | ill                      | illustrator |
| dc:creator or dc:contributor | art                      | artist      |
| dc:creator or dc:contributor | clr                      | colorist    |
| dc:creator or dc:contributor | nrt                      | narrator    |
```

File: streamer/parser/metadata.md [hunk 10]
```diff
@@ -111,52 +105,56 @@ The contributor object may also contain a `sortAs` string, used to sort the cont
...
+| dc:publisher   | pbl or \<empty\> or \<unknown\> | publisher   |
+| dc:contributor | trl                             | translator  |
+| dc:contributor | edt                             | editor      |
+| dc:contributor | ill                             | illustrator |
+| dc:contributor | art                             | artist      |
+| dc:contributor | clr                             | colorist    |
+| dc:contributor | nrt                             | narrator    |
+| media:narrator | nrt or \<empty\> or \<unknown\> | narrator    |
+| dc:contributor | \<empty\> or \<unknown\>        | contributor |
+
```
Comment (JayPanoz) [inline streamer/parser/metadata.md]:
```suggestion
| element                      | role                     | key         |
|------------------------------|--------------------------|-------------|
| dc:creator                   | \<empty\> or \<unknown\> | author      |
| dc:creator or dc:contributor | aut                      | author      |
| dc:contributor               | \<empty\> or \<unknown\> | contributor |
| dc:publisher                 | \<any\>                  | publisher   |
| dc:creator or dc:contributor | pbl                      | publisher   |
| dc:creator or dc:contributor | trl                      | translator  |
| dc:creator or dc:contributor | edt                      | editor      |
| dc:creator or dc:contributor | ill                      | illustrator |
| dc:creator or dc:contributor | art                      | artist      |
| dc:creator or dc:contributor | clr                      | colorist    |
| dc:creator or dc:contributor | nrt                      | narrator    |
| media:narrator               | \<any\>                  | narrator    |
```

File: streamer/parser/metadata.md (hunk 8 above)
Comment (JayPanoz) [reply streamer/parser/metadata.md]: Given the table has been updated, remove.

File: streamer/parser/metadata.md (hunk 7 above)
Comment (JayPanoz) [reply streamer/parser/metadata.md]: Given the table has been updated, remove.

File: streamer/parser/metadata.md [hunk 11]
```diff
@@ -111,52 +105,59 @@ The contributor object may also contain a `sortAs` string, used to sort the cont
...
+| dc:creator or dc:contributor | edt                      | editor      |
+| dc:creator or dc:contributor | ill                      | illustrator |
+| dc:creator or dc:contributor | art                      | artist      |
+| dc:creator or dc:contributor | clr                      | colorist    |
+| dc:creator or dc:contributor | nrt                      | narrator    |
 
 Where `opf:role` is the value of the attribute of the `<dc:element>`.
 
-The `name` of the contributor is the value of the element.
+In case of conflict, `opf:role` overrides the XML element used. For example, map a `<dc:publisher>` or `<dc:contributor>` element whose value for `opf:role` is `aut` to an author.
```
Comment (JayPanoz) [inline streamer/parser/metadata.md]:
```suggestion
As a rule of thumb, `opf:role` overrides `<dc:creator>` and `<dc:contributor>`. For example, map a `<dc:contributor>` element whose value for `opf:role` is `aut` to an author.
```

File: streamer/parser/metadata.md [hunk 12]
```diff
@@ -111,52 +105,59 @@ The contributor object may also contain a `sortAs` string, used to sort the cont
...
+| dc:creator or dc:contributor | ill                      | illustrator |
+| dc:creator or dc:contributor | art                      | artist      |
+| dc:creator or dc:contributor | clr                      | colorist    |
+| dc:creator or dc:contributor | nrt                      | narrator    |
+| media:narrator               | \<any\>                  | narrator    |
 
 Where `role` is the value of the refine whose `scheme` is a value of `marc:relators`.
 
-To handle the `name` of the contributor:
+In case of conflict, `role` overrides the XML element used. So, for example, map a `<dc:publisher>` or `<dc:contributor>` element whose value for `role` is `aut` to an author.
```
Comment (JayPanoz) [inline streamer/parser/metadata.md]:
```suggestion
As a rule of thumb, `role` overrides `<dc:creator>` and `<dc:contributor>`. For example, map a `<dc:contributor>` element whose value for `role` is `aut` to an author.
```

File: streamer/parser/metadata.md (hunk 11 above)
Comment (qnga) [reply streamer/parser/metadata.md]: I'm not sure about that. <dc:creator> doesn't imply a key by itself, and we decided that there is no clear distinction between `dc:creator` and `dc:contributor` too. So what is overridden?

Comment (JayPanoz) [reply streamer/parser/metadata.md]: We could use something else than `override`. I’m not sure how to rephrase this because I honestly can’t even tell whether I’ve been using those right or wrong for all those years at this point.

And, should we change this mapping then?

| element | opf:role | key |
|---------|----------|-----|
| dc:creator | \<empty\> or \<unknown\> | author |

It was my understanding this is how we decided to handle:

> It's not clear from the table how these elements are mapped when the opf:role attribute is missing. Is <dc:creator> ending up in authors and <dc:contributor> in contributors, like before?

But maybe it should be explored a little bit more – if it’s `<unknown>` should it be a `contributor` object with a `role`?

I mean I’ve just checked and a `<dc:creator>` without an `aut` refine passes EPUBCheck so we must have a default mapping for `<empty>`. But `<unknown>` I’m not so sure now.

Comment (qnga) [reply streamer/parser/metadata.md]: > In case of conflict, `opf:role` overrides the XML element used. For example, map a `<dc:publisher>` or `<dc:contributor>` element whose value for `opf:role` is `aut` to an author.

I think this sentence is simply useless with the updated table. All cases are covered.

It's hard to figure out a `<dc:creator>` with an unknown role. Maybe `abridger` or `interviewee` are enough important to be used with a `<dc:creator>` element, but should not be mapped to `author`.

Comment (JayPanoz) [reply streamer/parser/metadata.md]: Sounds reasonable, let’s remove those 2 paragraphs then. 

For documentation’s sake, if that can be useful in the future, a few references to [the EPUB 3.2 spec](https://www.w3.org/publishing/epub3/epub-packages.html#sec-opf-dccreator) (emphasis mine).

> The [DC11] creator element represents the name of a person, organization, etc. responsible for the creation of the content of the Rendition. The `role` property **can** be attached to the element to indicate the function the creator played in the creation of the content.

> The creator element SHOULD contain the name of the creator as the Author **intends it to be displayed to a user.**

There’s so much room for interpretation that I won’t firmly state what an `<empty>` role means but well at least it seems reasonable to me that in most cases, it will be the name on the cover and in the page header of a print book.

And thanks to everyone for their patience.

File: streamer/parser/metadata.md [hunk 13]
```diff
@@ -6,6 +6,17 @@ While the default context is very flexible in the way each metadata can be repre
...
+
+In many cases, the default context supports alternate representations of the same string in different scripts and languages by means of JSON-LD language maps.
+To fill such a map from an EPUB metadata element, proceed as follows:
+
+* Determine the language used in the content of the carrying element as defined in [the XML specification](https://www.w3.org/TR/xml/#sec-lang-tag),
+  i.e. check whether the carrying element has or inherits an `xml:lang` attribute. Otherwise, fallback to the primary language of the publication.
+* In the EPUB 3.x case, check if the element is refined by some `meta` elements that have or inherit an `xml:lang` attribute and whose property is `alternate-script`.
+  For each one, add to the map the corresponding language associated with the content of the `meta` element.
+
+
```
Comment (JayPanoz) [inline streamer/parser/metadata.md:19]: Question: should this be updated now that the decision to use `und` has been made?

Comment (qnga) [reply streamer/parser/metadata.md:19]: How would you update this? Specifying that as a last fallback `null` or `und` is used? (this is equivalent in the Kotlin code)
Or are you talking about removing 

> Otherwise, fallback to the primary language of the publication.

Unfortunately, as far as I know, no decision has been taken about that. Do you have an opinion? (see https://github.com/readium/webpub-manifest/issues/52#issuecomment-601625302)

Comment (JayPanoz) [reply streamer/parser/metadata.md:19]: Ah so during the review of the accessibility parsing doc, @llemeurfr mentioned the use of `und` here for `sortAs`: https://github.com/readium/architecture/pull/135#discussion_r417879475

So I assumed that resolution was made during on of the calls I missed and that this PR should perhaps be updated accordingly. In any case, there seems to be a discrepancy between PRs and issues right now.

Comment (qnga) [reply streamer/parser/metadata.md]: Okay, I drafted an update to take into account recent decisions.

Comment (JayPanoz) [reply streamer/parser/metadata.md]: Thanks!

**Comment by JayPanoz** [comment]
I guess this PR + #135 can be merged now.

Next step for these EPUB parsing docs will be, as discussed a month ago during the engineering call, re-organising the EPUB doc a little bit based on nav doc (landmarks), opf, etc.

**Comment by mickael-menu** [comment]
I'll merge it this Friday, unless someone vetoes it first.
